pkg/amber/provenance.go

```diff
@@ -71,19 +71,59 @@ func validateSLSAProvenanceJSON(provenanceFile []byte) error {
 }
 
 // ParseProvenanceFile reads a JSON file from a given path, and calls ParseStatementData on the
-// content of the file, if the read is successful.
+// content of the file, if the read is successful. It then sets all fields in ProvenanceIR to verify Amber provenances.
 // Returns an error if the file is not a valid provenance statement.
-func ParseProvenanceFile(path string) (*types.ValidatedProvenance, error) {
+func ParseProvenanceFile(path string) (*types.ProvenanceIR, error) {
 	statementBytes, err := os.ReadFile(path)
 	if err != nil {
 		return nil, fmt.Errorf("could not read the provenance file: %v", err)
 	}
-
 	if err := validateSLSAProvenanceJSON(statementBytes); err != nil {
 		return nil, err
 	}
 
-	return types.ParseStatementData(statementBytes)
+	provenanceIR, err := types.ParseStatementData(statementBytes)
+	if err != nil {
+		return nil, fmt.Errorf("could not parse the provenance bytes: %v", err)
+	}
+
+	if err := SetAmberProvenanceData(provenanceIR); err != nil {
```

Comment:
In general we want `ParseProvenanceFile` and `ParseStatementData` (or perhaps `ParseProvenanceData` considering the new change) to be only different in that one uses a file as input and the other uses a byte array as input. 

I suggest changing `SetAmberProvenanceData` to  take a byte array as input, first calling `types.ParseStatementData(statementBytes)` and then the rest. I also suggest renaming it to `ParseProvenanceData`. Then in `ParseProvenanceFile` as before, you only need to call one function (`ParseProvenanceData`) and no other logic.

Reply:
I am not sure `SetAmberProvenanceData` will work with the dependencies. 

I was quite struggling to get those right, but I might be missing something. 

I suggest I'll first move the definition of `ProvenanceIR` from `types` to `common` as in the comment below and then revisit this comment.

Reply:
Well yes, you don't want to introduce a dependency to `internal/common` in the files and packages in `pkg`. But if all these functions are tightly related to `ProvenanceIR` they should go into `common` too. 

But having a function `ParseProvenanceFile` in this file returning a data structure in `pkg` (e.g., `intoto.Statement`) gives a more convenient API.

Currently `types.ValidatedProvenance` is the glue between `internal` and `pkg`. The only thing that `types.ValidatedProvenance` adds to `intoto.Statement` is a validity check. If we only care about this validity in contexts where `ProvenanceIR` is used, then we can move the validation logic to a function that creates an instance of `ProvenanceIR`. Then we can remove `types.ValidatedProvenance`, and replace all its occurrences with `intoto.Statement`.

Reply:
> Currently types.ValidatedProvenance is the glue between internal and pkg.

Yes, you are right, and I think types.ValidatedProvenance being the glue is the key. 

I just tried to move `ProvenanceIR` to internal.common, which brings me in a full circle of where I started: needing both: `ProvenanceIR` and `ValidatedProvenance`.

Now after all this, I think we should keep both: `ValidatedProvenance` and `ProvenanceIR`. 

The main reason is:

We have to go from bytes to `ValidatedProvenance` to `ProvenanceIR`. The last step, from `ValidatedProvenace` to `ProvenanceIR` depends on `ValidatedProvenance`. 

So I do not think it a good idea to merge the two types, if we want to keep the flexibility, especially as

> replace all its occurrences with `intoto.Statement`

Yes, that is what we would need to do, but I feel we are loosing information and control over the type in this case.

To summarize: I think we should keep `ValidatedProvenance` and `ProvenanceIR`. 

However, I am not sure about the guarantees the two types give (and related their names).

Should `ProvenanceIR` hold the binary digest/name? More generally: do we, internally, only want to work on `ProvenanceIR`?

Reply:
> To summarize: I think we should keep `ValidatedProvenance` and `ProvenanceIR`.

Sounds good. 

> However, I am not sure about the guarantees the two types give (and related their names).
> 
> Should `ProvenanceIR` hold the binary digest/name? More generally: do we, internally, only want to work on `ProvenanceIR`?

I think it would be nice to do that. That would require converting `ValidatedProvenance` to `ProvenanceIR` as early as possible. Having separate fields in `ProvenanceIR` for digest and name makes sense to me, although IIRC, we won't have a reference value for name.